File: swift/planet-time/Sources/InterplanetTime/Orbital.swift
// Orbital.swift — Orbital mechanics ported from planet-time.js.

import Foundation

public extension InterplanetTime {

    // MARK: - Leap seconds / TT

    static func taiMinusUtc(_ utcMs: Int64) -> Int {
        var tai = 10
        for ls in leapSeconds {
            if utcMs >= ls.utcMs { tai = ls.delta } else { break }
        }
        return tai
    }

    /// Julian Ephemeris Day (TT) from UTC milliseconds.
    static func jde(_ utcMs: Int64) -> Double {
        let ttMs = Double(utcMs) + Double(taiMinusUtc(utcMs)) * 1000 + 32184
        return 2_440_587.5 + ttMs / 86_400_000.0
    }

    /// Julian centuries from J2000.0 (TT).
    static func jc(_ utcMs: Int64) -> Double {
        (jde(utcMs) - j2000JD) / 36_525.0
    }

    // MARK: - Kepler solver

    private static func keplerE(_ M: Double, _ e: Double) -> Double {
        var E = M
        for _ in 0..<50 {
            let dE = (M - E + e * sin(E)) / (1.0 - e * cos(E))
            E += dE
            if abs(dE) < 1e-12 { break }
        }
        return E
    }

    // MARK: - Heliocentric position

    struct HelioPos {
        public let x, y, r, lon: Double
    }

    static func helioPos(_ planet: String, _ utcMs: Int64) -> HelioPos {
        let el = orbitalElements[planet] ?? orbitalElements["earth"]!
        let T = jc(utcMs)
        let L = (el.L0 + el.dL * T).truncatingRemainder(dividingBy: 360.0)
        let om = el.om0
        let e  = el.e0
        let a  = el.a
        let Mrad = ((L - om + 360.0).truncatingRemainder(dividingBy: 360.0)) * .pi / 180.0
        let E = keplerE(Mrad, e)
        let nu = 2.0 * atan2(sqrt(1.0 + e) * sin(E / 2.0),
                             sqrt(1.0 - e) * cos(E / 2.0))
        let r   = a * (1.0 - e * cos(E))
        let lon = (om * .pi / 180.0 + nu + 2 * .pi).truncatingRemainder(dividingBy: 2 * .pi)
        return HelioPos(x: r * cos(lon), y: r * sin(lon), r: r, lon: lon)
    }

    // MARK: - Distance & light travel

    static func bodyDistanceAu(_ a: String, _ b: String, _ utcMs: Int64) -> Double {
        let pa = helioPos(a, utcMs), pb = helioPos(b, utcMs)
        let dx = pa.x - pb.x, dy = pa.y - pb.y
        return sqrt(dx*dx + dy*dy)
    }

    static func lightTravelSeconds(_ a: String, _ b: String, _ utcMs: Int64) -> Double {
        bodyDistanceAu(a, b, utcMs) * auSeconds
    }

    // MARK: - Line of sight

    struct LineOfSight {
        public let clear, blocked, degraded: Bool
        public let closestSunAu: Double?
        public let elongDeg: Double
    }

    static func checkLineOfSight(_ a: String, _ b: String, _ utcMs: Int64) -> LineOfSight {
        let pa = helioPos(a, utcMs), pb = helioPos(b, utcMs)
        let abx = pb.x - pa.x, aby = pb.y - pa.y
        let d2  = abx*abx + aby*aby
        if d2 < 1e-20 {
            return LineOfSight(clear: true, blocked: false, degraded: false,
                               closestSunAu: nil, elongDeg: 0)
        }
        let t = max(0.0, min(1.0, -(pa.x*abx + pa.y*aby) / d2))
        let cx = pa.x + t*abx, cy = pa.y + t*aby
        let closest = sqrt(cx*cx + cy*cy)
        let dotAB = abx*pa.x + aby*pa.y
        let abMag = sqrt(d2), aMag = sqrt(pa.x*pa.x + pa.y*pa.y)
        let cosEl = (aMag > 1e-10 && abMag > 1e-10) ? -dotAB / (abMag * aMag) : 0.0
        let elongDeg = acos(max(-1, min(1, cosEl))) * 180 / .pi
        let blocked  = closest < 0.1
        let degraded = !blocked && (closest < 0.25 || elongDeg < 5.0)
        return LineOfSight(clear: !blocked && !degraded, blocked: blocked, degraded: degraded,
                           closestSunAu: closest, elongDeg: elongDeg)
    }

    // MARK: - Lower-quartile light time

    static func lowerQuartileLightTime(_ a: String, _ b: String, refMs: Int64) -> Double {
        let yearMs = Int64(365) * earthDayMs
        let step   = yearMs / 360
        var samples = (0..<360).map { i in lightTravelSeconds(a, b, refMs + Int64(i) * step) }
        samples.sort()
        return samples[Int(Double(samples.count) * 0.25)]
    }
}

File: php/planet-time/src/InterplanetTime/Orbital.php
<?php
declare(strict_types=1);

namespace InterplanetTime;

/**
 * Orbital mechanics — ported from planet-time.js.
 *
 * All timestamps are UTC milliseconds since Unix epoch.
 * Distances in AU, angles in radians unless noted.
 */
final class Orbital
{
    private function __construct() {}

    // ── Leap seconds / TT ────────────────────────────────────────────────────

    public static function taiMinusUtc(int $utcMs): int
    {
        $ls = Constants::LEAP_SECONDS;
        $tai = 10;
        foreach ($ls as [$ts, $delta]) {
            if ($utcMs >= $ts) $tai = $delta;
            else break;
        }
        return $tai;
    }

    /** Julian Ephemeris Day (TT) from UTC milliseconds. */
    public static function jde(int $utcMs): float
    {
        $ttMs = $utcMs + self::taiMinusUtc($utcMs) * 1000 + 32184; // TT = TAI + 32.184s
        return 2440587.5 + $ttMs / 86400000.0;
    }

    /** Julian centuries from J2000.0 (TT). */
    public static function jc(int $utcMs): float
    {
        return (self::jde($utcMs) - Constants::J2000_JD) / 36525.0;
    }

    // ── Kepler solver ────────────────────────────────────────────────────────

    /** Solve Kepler's equation M = E - e*sin(E) via Newton-Raphson. */
    private static function keplerE(float $M, float $e): float
    {
        $E = $M;
        for ($i = 0; $i < 50; $i++) {
            $dE = ($M - $E + $e * sin($E)) / (1.0 - $e * cos($E));
            $E += $dE;
            if (abs($dE) < 1e-12) break;
        }
        return $E;
    }

    // ── Heliocentric position ─────────────────────────────────────────────────

    public static function helioPos(string $planet, int $utcMs): HelioPosResult
    {
        $elems = Constants::ORBITAL_ELEMENTS[$planet]
            ?? Constants::ORBITAL_ELEMENTS['earth'];

        $T  = self::jc($utcMs);
        $L  = fmod($elems['L0'] + $elems['dL'] * $T, 360.0);
        $om = $elems['om0'];
        $e  = $elems['e0'];
        $a  = $elems['a'];

        // Mean anomaly (deg → rad)
        $M = deg2rad(fmod($L - $om + 360.0, 360.0));
        $E = self::keplerE($M, $e);

        // True anomaly
        $nu = 2.0 * atan2(
            sqrt(1.0 + $e) * sin($E / 2.0),
            sqrt(1.0 - $e) * cos($E / 2.0)
        );

        // Heliocentric distance
        $r = $a * (1.0 - $e * cos($E));

        // Ecliptic longitude (radians)
        $lon = fmod(deg2rad($om) + $nu + 2.0 * M_PI, 2.0 * M_PI);

        return new HelioPosResult(
            x: $r * cos($lon),
            y: $r * sin($lon),
            r: $r,
            lon: $lon,
        );
    }

    // ── Distance & light travel ────────────────────────────────────────────────

    public static function bodyDistanceAu(string $a, string $b, int $utcMs): float
    {
        $pa = self::helioPos($a, $utcMs);
        $pb = self::helioPos($b, $utcMs);
        $dx = $pa->x - $pb->x;
        $dy = $pa->y - $pb->y;
        return sqrt($dx * $dx + $dy * $dy);
    }

    public static function lightTravelSeconds(string $a, string $b, int $utcMs): float
    {
        return self::bodyDistanceAu($a, $b, $utcMs) * Constants::AU_SECONDS;
    }

    // ── Line of sight ─────────────────────────────────────────────────────────

    public static function checkLineOfSight(string $a, string $b, int $utcMs): LineOfSightResult
    {
        $pa = self::helioPos($a, $utcMs);
        $pb = self::helioPos($b, $utcMs);

        // Vector from A to B
        $abx = $pb->x - $pa->x;
        $aby = $pb->y - $pa->y;
        $d2  = $abx * $abx + $aby * $aby;

        if ($d2 < 1e-20) {
            // Same body (Moon/Earth edge case)
            return new LineOfSightResult(
                clear: true, blocked: false, degraded: false,
                closestSunAu: null, elongDeg: 0.0,
            );
        }

        // Closest approach of Sun to line AB (t in [0,1])
        $t = max(0.0, min(1.0, -($pa->x * $abx + $pa->y * $aby) / $d2));
        $cx = $pa->x + $t * $abx;
        $cy = $pa->y + $t * $aby;
        $closestSunAu = sqrt($cx * $cx + $cy * $cy);

        // Solar elongation angle at observer A (degrees)
        $dotAB  = $abx * $pa->x + $aby * $pa->y;
        $abMag  = sqrt($d2);
        $aMag   = sqrt($pa->x ** 2 + $pa->y ** 2);
        $cos_el = ($aMag > 1e-10 && $abMag > 1e-10)
            ? -$dotAB / ($abMag * $aMag) : 0.0;
        $elongDeg = rad2deg(acos(max(-1.0, min(1.0, $cos_el))));

        $blocked  = $closestSunAu < 0.1;
        $degraded = !$blocked && ($closestSunAu < 0.25 || $elongDeg < 5.0);

        return new LineOfSightResult(
            clear: !$blocked && !$degraded,
            blocked: $blocked,
            degraded: $degraded,
            closestSunAu: $closestSunAu,
            elongDeg: $elongDeg,
        );
    }

    // ── Lower-quartile light time ──────────────────────────────────────────────

    /**
     * Sample one Earth year (360 samples) and return the lower-quartile
     * one-way light time in seconds (p25).
     */
    public static function lowerQuartileLightTime(string $a, string $b, int $refMs): float
    {
        $YEAR_MS = 365 * Constants::EARTH_DAY_MS;
        $STEP    = (int)($YEAR_MS / 360);
        $samples = [];
        for ($i = 0; $i < 360; $i++) {
            $samples[] = self::lightTravelSeconds($a, $b, $refMs + $i * $STEP);
        }
        sort($samples);
        return $samples[(int)(count($samples) * 0.25)];
    }
}
